### ISSUE TITLE
Scale Kubernetes clusters

### ISSUE THREAD

**Issue by mr-field** [UserOfficeProject/stfc-user-office-project#210]
Currently, both the dev and prod clusters are running with only 1 master node. This could cause problems in the future, as the master VM is a single point of failure: if it goes down, it compromises the whole cluster.

We should upgrade to a cluster with at least 3 master nodes both an dev and prod.

**Comment by TickleThePanda** [comment]
This is in "P2 (MVP 1: LSF)" but isn't really required until the last few weeks before the LSF deadline.

**Comment by mr-field** [comment]
In order to test this, I've asked SCD to increase our resource quota. The ticket number of the request is STFCCLOUD-378.

**Comment by mr-field** [comment]
Our resource quota on SCD Cloud has now been increased.

It seems that masters can't be added to a cluster after it's created, so we should test migrating dev to a new cluster with more masters. This will involve either pointing the Ingress IP to the new cluster, or asking DI to point the `devkubernetes.facilities.rl.ac.uk` record to the new cluster's ingress IP.

**Comment by mr-field** [comment]
> SCD suggested using the l2 flavour for worker nodes as there were more available, should we be doing that?

@TCMeldrum I looked at that, but even the smallest l2 VM (`l2.tiny`) is *huge* for what we need (8 VCPUs, 30 GB RAM, 200 GB disk). Just 5 of those nodes would take up our whole quota.

---

> What's the advantage of having 2 larger workers over 4 smaller workers?

@TickleThePanda It increases the maximum RAM that can be used by any service replica. 2GB seems pretty small for a production service, especially when you consider that some memory will be taken up by the OS and k8s services.

**Comment by mr-field** [comment]
At the moment, creating Magnum clusters on SCD Cloud always fails with this error: `Failed to get discovery url from 'https://discovery.etcd.io/new?size=3'`.

I've written to the cloud support team about this and they said they've been seeing this problem over the last week and are trying to fix it. They'll let me know when they have more information.

Until we can create a cluster again, this issue is blocked.

**Comment by mr-field** [comment]
I got an update from SCD today. They're still working on it, but it looks like they have a better idea of what's happening now:

> Hi Cosimo,
> 
> Update on this issue, it's taken us a while to track down but it appears to be a complex SSL encryption issue, we're working our way around the problem and apologise for the delay,
> 
> Thanks,
> Jacob

**Comment by mr-field** [comment]
Turns out my [workaround](https://github.com/isisbusapps/fba-reverse-proxy/pull/126) actually works! We can now create clusters again, as long as we remember to manually set their discovery URL to `https://devproxy-api.facilities.rl.ac.uk/etcd-discovery/<token>`. We have to manually generate the token in advance, with `curl https://discovery.etcd.io/new?size=<number of master nodes in the cluster>`. I didn't realise it was working earlier, because normally Magnum generates the token by itself when creating a cluster. However, it seems like it wants to be spoon-fed one if the discovery URL is assigned manually.

I've now created a new dev cluster with [the settings above](https://github.com/UserOfficeProject/stfc-user-office-project/issues/210#issuecomment-940188212), called `fba-cluster-dev`. The next step is to deploy our services to it and point the current dev floating IP to the new Ingress service.

**Comment by mr-field** [comment]
Dev is now running on the new cluster. To set up the new cluster, I just had to update the Jenkins `dev-kubeconfig` secret with the new cluster's Kubeconfig and run the deploy jobs again. To actually use the new cluster, I then had to disconnect the current Ingress floating IP from the old cluster and point it to the new one. The whole process only took a few minutes and there was practically no downtime.

We should do the same on prod next.

**Comment by mr-field** [comment]
The prod environment is running on a new cluster as well. The prod floating IP is now pointing at the new cluster and everything seems to be working OK.

I've replaced the old Kubeconfig files on the network shares with the ones for the new clusters. I've also added the SSH keypairs used to create the cluster resources to the same directory on the share.

While setting up the new cluster I has some trouble with timeouts when deploying the Ingress controller. The issue seems to be that STFC Cloud can take a while to allocate a load balancer to the Ingress, so after a while Helm thinks the deployment failed. We should extend the timeout or remove it so that we don't get this kind of error in the future.

**Comment by mr-field** [comment]
Would I be OK to tell STFC Cloud about our proxy for the discovery service? I'd make it clear that we don't offer any support for it, but it could be very useful for any other STFC users stuck in the same situation as us. The discovery service is only used during cluster creation, so it's very unlikely that our proxy would get a lot of traffic if more people knew about it. What do you think @TickleThePanda?

**Comment by TickleThePanda** [comment]
I'm happy with that. Let's hope they fix the issue in any case.

**Comment by mr-field** [comment]
I've shared the workaround with the Cloud team and opened https://github.com/isisbusapps/docker-orchestration/issues/200 to track the problem with the discovery service.

Both the dev and prod environment have now been switched to [newer, bigger clusters](https://github.com/UserOfficeProject/stfc-user-office-project/issues/210#issuecomment-940188212) and seem to work fine. The Kubeconfigs on the network drive have been updated and developers have been asked to copy them again.
